### ISSUE TITLE
How to set counter for nn.functional.interpolate?

### ISSUE THREAD

**Issue by fengyuentau** [Lyken17/pytorch-OpCounter#36]
In my model, `nn.functional.interpolate` is used, whose computation should be also calculated in my case.

`nn.functional.interpolate: count_upsample_bilinear` was appended in `register_hooks` in `profile.py`, and `count_upsample_bilinear` was implemented in `count_hooks.py`, but it did not work.

I also tried to print `m_type` in `add_hooks` in `profile`, but the output did not include `nn.functional.interpolate`. It seems that we cannot add hooks to some APIs like `nn.functional.interpolate` which work differently from APIs from `torch.nn` in terms of parameters and handler. I wonder if we can only wrap them into 3rd party modules to set FLOPs counters for them like the example in [README.md](https://github.com/Lyken17/pytorch-OpCounter/blob/master/README.md).

Is there any workarounds introducing minimum modification to the model definition code?

**Comment by fengyuentau** [comment]
After some testing, I found only the model definition in `__init__` (it doesn't matter whether the definition is used or not in `forward`) will be added hooks and profiled, which introduces a tricky problem for APIs like `nn.functional.interpolate` who needs the size of output to be passed when initializing, but the size of output can only be revealed as the tensor goes through the model.

**Comment by fengyuentau** [comment]
[A post](https://discuss.pytorch.org/t/calculating-flops-of-a-given-pytorch-model/3711/6?u=yuantao_feng) from PyTorch forum says, 
> model.apply(fn) combined with module.register_forward_hook(hook) allows for easy tracking of layers, but only works for nn.Module layers (conv, batchnorm, etc). This is effective for the majority of cases, but ***does not allow for tracking of functional calls, e.g. F.interpolate(...)***. Is there any way to detect functional calls in a forward pass?

It seems currently there is no solution for setting up FLOPs counter for `interpolate`.

**Comment by Lyken17** [comment]
That's true. Currently, THOP is based on module-level hooks. Can you try `nn.upsampling` instead?

**Comment by Lyken17** [comment]
I have implemented a version [count_hooks.py#L95](https://github.com/Lyken17/pytorch-OpCounter/blob/master/thop/count_hooks.py#L95).

**Comment by fengyuentau** [comment]
@Lyken17 Thanks for your implementation. The most difficult problem is I cannot specify the output size when the model is initializing since multi-scale testing is being popularly applied and the input size may not be fixed.

**Comment by Lyken17** [comment]
My current implementation is based on hooks during the forwarding, when the output size is determined.

**Comment by fengyuentau** [comment]
I realized that I can set the `scale_factor` instead of `size` for `nn.Upsample`. Also I made a mistake: I thought `nn.Upsample` had been deprecated, but the actually deprecated one should be `nn.functional.upsample`. Thanks for your help.

**Comment by fengyuentau** [comment]
A potential and tricky problem: If `scale_factor` is set for `nn.Upsample`, such as `scale_factor=2`, and the height/width is odd, the upsampled size would not match the original size. For example, given an input of size `41x41`, the input is downsampled by Conv2d which outputs feature maps of size `21x21`. Upsampling a `21x21` tensor by `scale_factor=2` outputs a tensor of size `42x42` which does not match with the original one.

However, to register a forward hook for `nn.Upsample`, `nn.Upsample` must be defined in the `__init__` function of the model. It becomes tricky if the input size is not fixed and you can only set `scale_factor` to instantiate a `nn.Upsample` in `__init__`.

**Comment by fengyuentau** [comment]
A workaround: set the input size to be the multiple of the downsample factor. For example, ResNet downsamples an input by a factor of 32, meaning if one wants to apply upsample to the tensor of size 32 times smaller, the input size must be the multiple of 32 to keep the sizes of tensors matched. Or if the upsample is applied to the tensor of size 16 times smaller, then the input size should be the multiple of 16.
